### PR TITLE
backport: grpc: add automatic retry support to all services

### PR DESCRIPTION
**Pull request by ggilmore**
This is a manual backport of all the following PRs into a single commit for the 5.2 branch:

- https://github.com/sourcegraph/sourcegraph/pull/59095
- https://github.com/sourcegraph/sourcegraph/pull/59140
- https://github.com/sourcegraph/sourcegraph/pull/59145
- https://github.com/sourcegraph/sourcegraph/pull/59146
- https://github.com/sourcegraph/sourcegraph/pull/59191
- https://github.com/sourcegraph/sourcegraph/pull/59107
- https://github.com/sourcegraph/sourcegraph/pull/59110
- https://github.com/sourcegraph/sourcegraph/pull/59111
- https://github.com/sourcegraph/sourcegraph/pull/59130
- https://github.com/sourcegraph/sourcegraph/pull/59133
- https://github.com/sourcegraph/sourcegraph/pull/59136

Collectively, these PRs all implement adding automatic retry support for idempotent gRPC methods to the 5.2 branch. Creating stacked PRs for these for the 5.2 backport would have taken ages with the need for manual approvals for each one, so I collected them all in one PR for convenience. 

Some services had methods that only exist on the 5.2 branch - we've deleted / changed them on `main`. I have called these unique methods out as PR comments.

## Test plan

CI for all referenced PRs is sufficient.

---

I ran the following search query on a local instance that had one instance of zoekt-webserver disabled (sg start --except zoekt-web-1): https://sourcegraph.test:3443/search?q=context%3Aglobal+test&patternType=standard&sm=1&trace=1&groupBy=repo

This produced a trace showing that the client did attempt to retry:

Screenshot 2024-01-09 at 1 49 31 PM 1

### CODE REVIEW

File: internal/gitserver/retry.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	return r.base.RepoCloneProgress(ctx, in, opts...)
+}
+
+func (r *automaticRetryClient) RepoUpdate(ctx context.Context, in *proto.RepoUpdateRequest, opts ...grpc.CallOption) (*proto.RepoUpdateResponse, error) {
+	opts = append(defaults.RetryPolicy, opts...)
+	return r.base.RepoUpdate(ctx, in, opts...)
+}
+
+func (r *automaticRetryClient) ReposStats(ctx context.Context, in *proto.ReposStatsRequest, opts ...grpc.CallOption) (*proto.ReposStatsResponse, error) {
+	opts = append(defaults.RetryPolicy, opts...)
```
Comment (ggilmore) [inline internal/gitserver/retry.go:105]: cc @sourcegraph/source 

Note that RepoStats doesn't exist on 5.3, only on 5.2. I thought this method had no side effects, but cc'ing y'all to make sure.

File: internal/gitserver/v1/gitserver.proto [hunk 2]
```diff
@@ -8,22 +8,46 @@ import "google/protobuf/timestamp.proto";
...
+  }
   rpc RepoDelete(RepoDeleteRequest) returns (RepoDeleteResponse) {}
-  rpc RepoUpdate(RepoUpdateRequest) returns (RepoUpdateResponse) {}
+  rpc RepoUpdate(RepoUpdateRequest) returns (RepoUpdateResponse) {
+    option idempotency_level = IDEMPOTENT;
+  }
   // TODO: Remove this endpoint after 5.2, it is deprecated.
-  rpc ReposStats(ReposStatsRequest) returns (ReposStatsResponse) {}
+  rpc ReposStats(ReposStatsRequest) returns (ReposStatsResponse) {
+    option idempotency_level = IDEMPOTENT;
```
Comment (ggilmore) [inline internal/gitserver/v1/gitserver.proto:49]: cc @sourcegraph/source 

Note that RepoStats doesn't exist on 5.3, only on 5.2. I thought this method had no side effects, but cc'ing y'all to make sure.

File: internal/repoupdater/v1/repoupdater.proto [hunk 3]
```diff
@@ -8,19 +8,33 @@ option go_package = "github.com/sourcegraph/sourcegraph/internal/repoupdater/v1"
...
-  rpc EnqueueChangesetSync(EnqueueChangesetSyncRequest) returns (EnqueueChangesetSyncResponse);
+  rpc EnqueueRepoUpdate(EnqueueRepoUpdateRequest) returns (EnqueueRepoUpdateResponse) {
+    option idempotency_level = IDEMPOTENT;
+  }
+  rpc EnqueueChangesetSync(EnqueueChangesetSyncRequest) returns (EnqueueChangesetSyncResponse) {
+    option idempotency_level = IDEMPOTENT;
+  }
   // SyncExternalService requests the given external service to be synced.
-  rpc SyncExternalService(SyncExternalServiceRequest) returns (SyncExternalServiceResponse);
+  rpc SyncExternalService(SyncExternalServiceRequest) returns (SyncExternalServiceResponse) {
```
Comment (ggilmore) [inline internal/repoupdater/v1/repoupdater.proto:27]: cc @sourcegraph/source 

Note that `main` has gotten rid of the SyncExternalService, ExternalServiceNamespace, and ExternalServiceRepositories methods - these only exit on the 5.2 branch.  I made a judgement call that these were all idempotent methods - but please correct me on this.

File: internal/symbols/v1/symbols.proto [hunk 4]
```diff
@@ -7,11 +7,21 @@ import "google/protobuf/duration.proto";
...
 service SymbolsService {
-  rpc Search(SearchRequest) returns (SearchResponse) {}
-  rpc LocalCodeIntel(LocalCodeIntelRequest) returns (stream LocalCodeIntelResponse) {}
-  rpc ListLanguages(ListLanguagesRequest) returns (ListLanguagesResponse) {}
-  rpc SymbolInfo(SymbolInfoRequest) returns (SymbolInfoResponse) {}
-  rpc Healthz(HealthzRequest) returns (HealthzResponse) {}
+  rpc Search(SearchRequest) returns (SearchResponse) {
+    option idempotency_level = NO_SIDE_EFFECTS;
+  }
+  rpc LocalCodeIntel(LocalCodeIntelRequest) returns (stream LocalCodeIntelResponse) {
```
Comment (ggilmore) [inline internal/symbols/v1/symbols.proto:13]: cc @sourcegraph/team-graph 

Note that `main` has gotten rid of the `LocalCodeIntel` method, it only exist on the 5.2 branch. I though this method seemed like it had no side effects, so it should be safe to retry. Please feel free to correct me on this though.

File: internal/symbols/retry.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+type automaticRetryClient struct {
+	base proto.SymbolsServiceClient
+}
+
+func (a *automaticRetryClient) Search(ctx context.Context, in *proto.SearchRequest, opts ...grpc.CallOption) (*proto.SearchResponse, error) {
+	opts = append(defaults.RetryPolicy, opts...)
+	return a.base.Search(ctx, in, opts...)
+}
+
+func (a *automaticRetryClient) LocalCodeIntel(ctx context.Context, in *proto.LocalCodeIntelRequest, opts ...grpc.CallOption) (proto.SymbolsService_LocalCodeIntelClient, error) {
```
Comment (ggilmore) [inline internal/symbols/retry.go:28]: cc @sourcegraph/team-graph 

Note that `main` has gotten rid of the `LocalCodeIntel` method, it only exist on the 5.2 branch. I though this method seemed like it had no side effects, so it should be safe to retry. Please feel free to correct me on this though.

**Comment by ggilmore** [comment]
@sourcegraph/release-guild Do I need to pay attention to the lighthouse errors? https://buildkite.com/sourcegraph/aspect-experimental/builds/4460

File: internal/gitserver/retry.go (hunk 1 above)
Comment (eseliger) [reply internal/gitserver/retry.go:105]: yes that's correct!

File: internal/repoupdater/v1/repoupdater.proto (hunk 3 above)
Comment (eseliger) [reply internal/repoupdater/v1/repoupdater.proto:27]: Should be fine 👍

**Review by eseliger — APPROVED** [review]
🚀

**Comment by ggilmore** [comment]
Hello @sourcegraph/release-guild - Could y'all stamp this PR and merge into the patch release? Thanks!

File: internal/symbols/v1/symbols.proto (hunk 4 above)
Comment (varungandhi-src) [reply internal/symbols/v1/symbols.proto:13]: Yes, it should be OK to retry.

**Pull request merged**